interfaces/prompting/requestprompts/requestprompts.go

```diff
@@ -0,0 +1,373 @@
+	_, err = io.ReadFull(f, idBuf[:])
+	if err != nil {
+		return fmt.Errorf("cannot read maximum prompt ID: %w", err)
+	}
+
+	maxID, err := strconv.ParseUint(string(idBuf[:]), 16, 64)
+	if err != nil {
+		return fmt.Errorf("cannot parse maximum prompt ID: %w", err)
+	}
+
+	pdb.maxID = maxID
+	return nil
+}
+
+// nextID advances the internal monotonically-increasing maxID integer.
+//
+// The caller must ensure that the prompt DB mutex is held.
+func (pdb *PromptDB) nextID() string {
+	pdb.maxID++
+	idStr := pdb.maxIDString()
+	osutil.AtomicWriteFile(pdb.maxIDPath, []byte(idStr), 0600, 0)
+	return idStr
+}
+
+// maxIDString returns a 16-character string corresponding to the current maxID.
+// The ID string is the max ID in hexadecimal, padded by leading zeroes.
+func (pdb *PromptDB) maxIDString() string {
+	return fmt.Sprintf("%016X", pdb.maxID)
+}
+
+// AddOrMerge checks if the given prompt contents are identical to an existing
+// prompt and, if so, merges with it by adding the given listenerReq to it.
+// Otherwise, adds a new prompt with the given contents to the prompt DB.
+//
+// If the prompt was merged with an identical existing prompt, returns the
+// existing prompt and true, indicating it was merged. If a new prompt was
+// added, returns the new prompt and false, indicating the prompt was not
+// merged.
+//
+// The caller must ensure that the given permissions are in the order in which
+// they appear in the available permissions list for the given interface.
+func (pdb *PromptDB) AddOrMerge(user uint32, snap string, iface string, path string, permissions []string, listenerReq *listener.Request) (*Prompt, bool) {
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+	userEntry, ok := pdb.perUser[user]
+	if !ok {
+		pdb.perUser[user] = &userPromptDB{
+			ByID: make(map[string]*Prompt),
+		}
+		userEntry = pdb.perUser[user]
+	}
+
+	availablePermissions, _ := prompting.AvailablePermissions(iface)
+	// Error should be impossible, since caller has already validated that iface
+	// is valid, and tests check that all valid interfaces have valid available
+	// permissions returned by AvailablePermissions.
+
+	constraints := &promptConstraints{
+		Path:                 path,
+		Permissions:          permissions,
+		AvailablePermissions: availablePermissions,
+	}
+
+	// Search for an identical existing prompt, merge if found
+	for _, prompt := range userEntry.ByID {
+		if prompt.Snap == snap && prompt.Interface == iface && prompt.Constraints.equals(constraints) {
+			prompt.listenerReqs = append(prompt.listenerReqs, listenerReq)
+			// Although the prompt itself has not changed, re-record a notice
+			// to re-notify clients to respond to this request. A client may
+			// have replied with a malformed response and not retried after
+			// receiving the error, so this notice encourages it to try again
+			// if the user retries the operation.
+			pdb.notifyPrompt(user, prompt.ID)
+			return prompt, true
+		}
+	}
+
+	id := pdb.nextID()
```

Comment:
could the ID be derived from request data? this would save the trouble of all the bookkeeping with an auxiliary file

Reply:
The request does have an opaque ID from the kernel, but we don't have guarantees that the kernel won't re-use these after a particular listener request is satisfied. So I'm not sure we can do better than bookkeeping our own monotonically-increasing ID. This might also make integration testing easier in the future ("expect prompt with id 2 with these details: ...").